File: bindings/js/examples/search-sudoku.mjs
import { Model } from "../search.mjs";

const model = new Model("4x4 Sudoku");
const cell = model.intGrid("cell", 4, 4, 1, 4);
for (const row of cell) model.allDifferent(row);
for (let column = 0; column < 4; column++) {
  model.allDifferent(cell.map((row) => row[column]));
}
for (const boxRow of [0, 2]) {
  for (const boxColumn of [0, 2]) {
    model.allDifferent(
      [0, 1].flatMap((dr) =>
        [0, 1].map((dc) => cell[boxRow + dr][boxColumn + dc]),
      ),
    );
  }
}
for (const [row, column, value] of [
  [0, 0, 1],
  [0, 3, 4],
  [1, 1, 4],
  [2, 2, 4],
  [3, 0, 4],
  [3, 3, 1],
]) {
  model.add(`${cell[row][column]} == ${value}`);
}
const solution = model.solve().requireSolution();
for (const row of cell)
  console.log(row.map((value) => solution.get(value)).join(" "));

File: bindings/python/examples/search_sudoku.py
"""Solve a 4x4 Sudoku with AY Search."""

from aysearch import Model


model = Model("4x4 Sudoku")
cell = model.int_grid("cell", 4, 4, 1, 4)

for row in cell:
    model.all_different(row)
for column in range(4):
    model.all_different(cell[row][column] for row in range(4))
for box_row in (0, 2):
    for box_column in (0, 2):
        model.all_different(
            cell[row][column]
            for row in range(box_row, box_row + 2)
            for column in range(box_column, box_column + 2)
        )

for row, column, value in (
    (0, 0, 1),
    (0, 3, 4),
    (1, 1, 4),
    (2, 2, 4),
    (3, 0, 4),
    (3, 3, 1),
):
    model.add(f"{cell[row][column]} == {value}")

solution = model.solve().require_solution()
for row in cell:
    print(" ".join(str(solution[value]) for value in row))
